File: Haskell/Cliente.hs
module Cliente(
    Cliente(..),
    geraCliente,
    formataCliente,
    clienteReserva,
    clienteCpf,
    clientePlaca,
    clienteStatus,
    clienteVeiculo,
    clienteVisitas,
    clienteVaga,
    clienteMostraCliente,
    clienteAtualizaValor,
    salvaCliente,
    recuperaCliente,
    lerArquivo,
    atualizaCliente
)

where

import  Data.List.Split
import System.Directory

data Cliente = Cliente{cpf:: String, placa::String, veiculo::String, status::String, valor::Float, vaga::Int, hora::Int, minuto::Int,visitas::Int, reserva::Bool} deriving (Show)

clienteAlteraVaga::Int->Cliente->Cliente
clienteAlteraVaga vaga (Cliente a b c d e f g h i j) = ((Cliente a b c d e vaga g h i j))

clienteReserva::Cliente->Bool
clienteReserva (Cliente _ _ _ _ _ _ _ _ _ reserva) = reserva

clienteVisitas::Cliente->Int
clienteVisitas (Cliente _ _ _ _ _ _ _ _ visitas _) = visitas

clientePlaca::Cliente->String
clientePlaca (Cliente _ placa _ _ _ _ _ _ _ _) = placa

clienteCpf::Cliente->String
clienteCpf (Cliente cpf _ _ _ _ _ _ _ _ _) = cpf

clienteStatus::Cliente->String
clienteStatus (Cliente _ _ _ status _ _ _ _ _ _) = status

clienteVeiculo::Cliente->String
clienteVeiculo (Cliente _ _ veiculo _ _ _ _ _ _ _) =  veiculo

clienteVaga::Cliente->Int
clienteVaga (Cliente _ _ _ _ _ vaga _ _ _ _) = vaga

clienteAtualizaValor::Cliente->Float->Cliente
clienteAtualizaValor (Cliente a b c d e f g h i j) novo = (Cliente a b c d novo f g h i j)

clienteMostraCliente::Cliente->String
clienteMostraCliente (Cliente a b c d e f g h i j) = "\nCPF: " ++ a ++ "\nPlaca: " ++ b ++ "\nVeiculo: " ++ c ++ "\nStatus: " ++ d ++ "\nValor: " ++ (show e) ++ "\nVaga: " ++ (show f) ++ "\nHora: " ++ (show g) ++ ":" ++ (show h) ++ "\nVisitas: " ++ (show i)

stringToBool::String->Bool
stringToBool "True" = True
stringToBool "False" = False

geraCliente:: String->Cliente
geraCliente cliente  =
    if cliente /= "" then(Cliente (dados !! 0) (dados !! 1) (dados !! 2) (dados !! 3) (read (dados !! 4)::Float) (read(dados !! 5)::Int) (read(dados !! 6)::Int) (read(dados !! 7)::Int) (read(dados !! 8)::Int) (stringToBool (dados !! 9)))
    else (Cliente "" "" "" "" 0 0 0 0 0 False)
    where
        dados = splitOn "-" cliente

formataCliente:: Cliente->String
formataCliente (Cliente a b c d e f g h i j) = a ++ "-" ++ b ++ "-" ++ c ++ "-" ++ d ++ "-" ++ (show e) ++ "-" ++ (show f) ++  "-" ++ (show g) ++ "-" ++ (show h) ++ "-" ++ (show i) ++ "-" ++ (show j) ++ "\n"

isCliente:: [String]->String->String
isCliente [] key = ""
isCliente (x:xs) key =
    if (splitOn "-" x !! 0) == key then x
    else isCliente xs key

recuperaCliente:: String->String-> String
recuperaCliente clientes cpf = do
    encontrado
    where
        encontrado = isCliente (splitOn "\n" clientes) cpf

lerArquivo::IO String
lerArquivo = readFile "dados/clientes.txt"

salvaCliente::Cliente->IO()
salvaCliente cliente  = do
    clientes <- lerArquivo
    let flag = recuperaCliente clientes (clienteCpf cliente)
    if flag /= "" then putStrLn $ "Cliente já existe"
    else do
        appendFile "dados/clientes.txt" $ formataCliente cliente

atualizaClienteAux::String->[String]->String->[String]
atualizaClienteAux cpf [] cliente = []
atualizaClienteAux cpf (x:xs) cliente =
    if (antigo !! 0) == (novo !! 0) then (cliente):atualizaClienteAux cpf xs cliente
    else x:atualizaClienteAux cpf xs cliente
    where
        antigo = splitOn "-" x
        novo = splitOn "-" cliente

atualizaCliente::String->String->IO()
atualizaCliente cpf cliente= do
    clientes <- lerArquivo
    let clienteAtualizado = atualizaClienteAux cpf (splitOn "\n" clientes) cliente
    removeFile "dados/clientes.txt"
    writeFile "dados/clientes.txt" $ (unlines clienteAtualizado)

File: prolog/cliente.pl
:- module(cliente, [formata_cliente/2,adiciona_cliente/1,atualiza_cliente/2,imprime_cliente/1,recupera_cliente/2,get_cpf/2,get_hora/2,get_vaga/2,get_placa/2,get_status/2,get_reserva/2,get_veiculo/2,get_visitas/2]).
:- use_module(library(readutil)).

get_cpf(Cliente,Cpf) :-
    nth0(0,Cliente,Cpf).
get_placa(Cliente,Placa) :-
    nth0(1,Cliente,Placa).
get_veiculo(Cliente,Veiculo) :-
    nth0(2,Cliente,Veiculo).
get_vaga(Cliente,Vaga) :-
    nth0(5,Cliente,Vaga).
get_reserva(Cliente,Reserva) :-
    nth0(8,Cliente,Reserva).
get_status(Cliente,Status) :-
    nth0(3,Cliente,Status).
get_visitas(Cliente,Visitas) :-
    nth0(7,Cliente,Visitas).
get_hora(Cliente,Hora) :-
    nth0(6,Cliente,Hora).

imprime_cliente(L) :-
    nth0(0,L,Cpf),nth0(1,L,Placa),nth0(2,L,Veiculo),nth0(3,L,Status),
    nth0(4,L,Valor),nth0(5,L,Vaga),nth0(6,L,Hora),nth0(7,L,Visitas),
    format(string(S),'CPF:~s\nPlaca:~s\nVeiculo:~s\nStatus:~s\nValor:~d\nVaga:~d\nHora:~d\nVisitas:~d',[Cpf,Placa,Veiculo,Status,Valor,Vaga,Hora,Visitas]),
    write(S).

formata_cliente(Out,L) :-
    format(string(Out),'~s-~s-~s-~s-~d-~s-~d-~s-~s',L).

adiciona_cliente(C) :-
    open('dados/clientes.txt',append,File),
    formata_cliente(S,C),
    writeln(File,S),
    close(File).

target([H|T],Flag) :- H == Flag.

encontrado([],_,[]).
encontrado([X|Xs],Cpf,Cliente) :-
    split_string(X,"-","",X1),
    target(X1,Cpf),
    Cliente = X1,!.
encontrado([X|Xs],Cpf,Cliente):- encontrado(Xs,Cpf,Cliente).

recupera_cliente(Cpf,Cliente) :-
    ler_arquivo(Clientes),
    encontrado(Clientes,Cpf,Cliente).


atualiza([],_,_,[]).
atualiza([X|Xs],Cpf,NovoCliente,[NovoCliente1|Resultado]) :-
    split_string(X,"-","",X1),
    target(X1,Cpf),
    formata_cliente(NovoCliente1,NovoCliente),
    atualiza(Xs,Cpf,NovoCliente,Resultado).
atualiza([X|Xs],Cpf,NovoCliente,[X|Resultado]) :- atualiza(Xs,Cpf,NovoCliente,Resultado).

atualiza_cliente(Cpf,NovoCliente) :-
    ler_arquivo(Clientes),
    atualiza(Clientes,Cpf,NovoCliente,Atualizado),
    atomic_list_concat(Atualizado,"\n",S),
    delete_file('dados/clientes.txt'),
    open('dados/clientes.txt',write,File),
    write(File,S),
    close(File).

ler_arquivo(Result) :-
    open('dados/clientes.txt',read,Str),
    read_stream_to_codes(Str,Clientes),
    atom_string(Clientes,Clientes1),
    split_string(Clientes1,"\n","",Result).
